File: tests/main/capability.test.ts
import { describe, it, expect, beforeEach, vi } from 'vitest'
import { EventEmitter } from 'events'
import { parseMessage } from '../../src/main/irc/parser'
import { dispatchMessage } from '../../src/main/irc/handlers/registry'
import { ConnectionState } from '../../src/main/irc/state'

// Import handler modules to register them
import '../../src/main/irc/handlers/registration'
import '../../src/main/irc/handlers/channel'
import '../../src/main/irc/handlers/message'
import '../../src/main/irc/handlers/user'
import '../../src/main/irc/handlers/error'
import '../../src/main/irc/capability'
import '../../src/main/irc/sasl'

function createMockClient() {
  const state = new ConnectionState()
  state.nick = 'TestUser'
  state.capNegotiating = true

  const events = new EventEmitter()
  const sentLines: string[] = []
  const connection = {
    send: (...args: string[]) => sentLines.push(args.join(' ')),
    sendRaw: (line: string) => sentLines.push(line)
  }

  const config = {
    nick: 'TestUser',
    autoJoin: [],
    saslMechanism: null,
    saslUsername: null,
    saslPassword: null,
    password: null
  }

  return {
    client: { state, events, connection, config } as any,
    events,
    state,
    sentLines
  }
}

describe('CAP Negotiation', () => {
  it('acts on a multi-line ACK only when its last line arrives', () => {
    const { client, state, sentLines } = createMockClient()
    dispatchMessage(client, parseMessage(':server CAP * LS :multi-prefix server-time'))
    expect(state.pendingCapRequests).toBe(1)

    // "a client MUST NOT change capabilities until the last ACK of the set"
    dispatchMessage(client, parseMessage(':server CAP * ACK * :multi-prefix'))
    expect(state.capabilities.has('multi-prefix')).toBe(true)
    expect(sentLines).not.toContain('CAP END')

    dispatchMessage(client, parseMessage(':server CAP * ACK :server-time'))
    expect(state.capabilities.has('server-time')).toBe(true)
    expect(sentLines).toContain('CAP END')
  })

  it('keeps the CAP LS of two networks apart', () => {
    // Every launch with more than one network is this: two servers listing
    // what they offer at the same moment, each over several lines
    const a = createMockClient()
    const b = createMockClient()

    dispatchMessage(a.client, parseMessage(':server-a CAP * LS * :multi-prefix'))
    dispatchMessage(b.client, parseMessage(':server-b CAP * LS * :sasl'))
    dispatchMessage(a.client, parseMessage(':server-a CAP * LS :server-time'))

    const askedA = a.sentLines.find((l) => l.startsWith('CAP REQ'))!
    expect(askedA).toContain('multi-prefix')
    expect(askedA).toContain('server-time')
    expect(askedA).not.toContain('sasl')
    expect(b.sentLines.find((l) => l.startsWith('CAP REQ'))).toBeUndefined()

    dispatchMessage(b.client, parseMessage(':server-b CAP * LS :echo-message'))
    const askedB = b.sentLines.find((l) => l.startsWith('CAP REQ'))!
    expect(askedB).toContain('sasl')
    expect(askedB).toContain('echo-message')
    expect(askedB).not.toContain('multi-prefix')
  })

  it('parses CAP LS and requests known caps', () => {
    const { client, sentLines } = createMockClient()

    dispatchMessage(
      client,
      parseMessage(':server CAP * LS :multi-prefix sasl message-tags server-time echo-message')
    )

    // Should have sent a CAP REQ for the known caps
    const reqLine = sentLines.find((l) => l.startsWith('CAP REQ'))
    expect(reqLine).toBeDefined()
    expect(reqLine).toContain('multi-prefix')
    expect(reqLine).toContain('sasl')
    expect(reqLine).toContain('message-tags')
    expect(reqLine).toContain('server-time')
    expect(reqLine).toContain('echo-message')
  })

  it('handles CAP LS with values', () => {
    const { client, state } = createMockClient()

    dispatchMessage(client, parseMessage(':server CAP * LS :sasl=PLAIN,EXTERNAL multi-prefix'))

    expect(state.availableCapabilities.get('sasl')).toBe('PLAIN,EXTERNAL')
    expect(state.availableCapabilities.get('multi-prefix')).toBeNull()
  })

  it('handles multi-line CAP LS', () => {
    const { client, sentLines } = createMockClient()

    // First line with * continuation marker
    dispatchMessage(client, parseMessage(':server CAP * LS * :multi-prefix sasl message-tags'))

    // Should NOT have sent REQ yet
    expect(sentLines.filter((l) => l.startsWith('CAP REQ')).length).toBe(0)

    // Second (final) line
    dispatchMessage(client, parseMessage(':server CAP * LS :server-time echo-message'))

    // NOW it should request
    const reqLine = sentLines.find((l) => l.startsWith('CAP REQ'))
    expect(reqLine).toBeDefined()
  })

  it('handles CAP ACK and sends CAP END (no SASL)', () => {
    const { client, sentLines, state } = createMockClient()

    dispatchMessage(
      client,
      parseMessage(':server CAP * ACK :multi-prefix message-tags server-time')
    )

    expect(state.capabilities.has('multi-prefix')).toBe(true)
    expect(state.capabilities.has('message-tags')).toBe(true)
    expect(state.capabilities.has('server-time')).toBe(true)
    expect(sentLines).toContain('CAP END')
    expect(state.capNegotiating).toBe(false)
  })

  it('handles CAP ACK with SASL — does NOT send CAP END yet', () => {
    const { client, sentLines, state } = createMockClient()
    client.config.saslMechanism = 'PLAIN'
    client.config.saslPassword = 'hunter2'

    dispatchMessage(client, parseMessage(':server CAP * ACK :sasl multi-prefix'))

    expect(state.capabilities.has('sasl')).toBe(true)
    expect(sentLines).toContain('AUTHENTICATE PLAIN')
    expect(sentLines).not.toContain('CAP END')
    // CAP END should be sent after SASL completes
  })

  /**
   * The one that started this.
   *
   * Somebody fills in an account name and a password and expects to be logged
   * in. The phone worked it out and authenticated with PLAIN; the desktop
   * required a mechanism to have been picked from a menu and otherwise
   * connected as a stranger, saying nothing. Same config, two behaviours.
   */
  it('logs in from a username and a password with no mechanism chosen', () => {
    const { client, sentLines } = createMockClient()
    client.config.saslMechanism = null
    client.config.saslUsername = 'kara'
    client.config.saslPassword = 'hunter2'

    dispatchMessage(client, parseMessage(':server CAP * ACK :sasl'))

    expect(sentLines).toContain('AUTHENTICATE PLAIN')
    expect(sentLines).not.toContain('CAP END')
  })

  /** And a mechanism with nothing to send says so rather than trying anyway */
  it('refuses a mechanism it has no password for, out loud', () => {
    const { client, sentLines, state } = createMockClient()
    client.config.saslMechanism = 'PLAIN'
    client.config.saslPassword = null

    const errors: { message: string }[] = []
    client.events.on('error', (err: { message: string }) => errors.push(err))

    dispatchMessage(client, parseMessage(':server CAP * ACK :sasl'))

    expect(sentLines).not.toContain('AUTHENTICATE PLAIN')
    expect(sentLines).toContain('CAP END')
    expect(state.capNegotiating).toBe(false)
    expect(errors.some((e) => /needs a password/i.test(e.message))).toBe(true)
  })

  it('handles CAP NAK — sends CAP END', () => {
    const { client, sentLines, state } = createMockClient()

    dispatchMessage(client, parseMessage(':server CAP * NAK :some-unknown-cap'))

    expect(sentLines).toContain('CAP END')
    expect(state.capNegotiating).toBe(false)
  })

  it('handles CAP NEW — requests new known caps', () => {
    const { client, sentLines } = createMockClient()

    dispatchMessage(client, parseMessage(':server CAP * NEW :away-notify account-notify'))

    const reqLine = sentLines.find((l) => l.startsWith('CAP REQ'))
    expect(reqLine).toBeDefined()
    expect(reqLine).toContain('away-notify')
    expect(reqLine).toContain('account-notify')
  })

  it('handles CAP DEL — removes caps from state', () => {
    const { client, state } = createMockClient()
    state.capabilities.add('away-notify')
    state.availableCapabilities.set('away-notify', null)

    dispatchMessage(client, parseMessage(':server CAP * DEL :away-notify'))

    expect(state.capabilities.has('away-notify')).toBe(false)
    expect(state.availableCapabilities.has('away-notify')).toBe(false)
  })

  it('sends CAP END with no requestable caps', () => {
    const { client, sentLines, state } = createMockClient()

    dispatchMessage(client, parseMessage(':server CAP * LS :some-unknown-cap another-unknown'))

    expect(sentLines).toContain('CAP END')
    expect(state.capNegotiating).toBe(false)
  })
})

describe('SASL Authentication', () => {
  it('handles SASL PLAIN authentication flow', () => {
    const { client, sentLines } = createMockClient()
    client.config.saslMechanism = 'PLAIN'
    client.config.saslUsername = 'myuser'
    client.config.saslPassword = 'mypass'

    // Server says ready
    dispatchMessage(client, parseMessage('AUTHENTICATE +'))

    // Should have sent base64-encoded credentials
    const authLine = sentLines.find((l) => l.startsWith('AUTHENTICATE') && l !== 'AUTHENTICATE +')
    expect(authLine).toBeDefined()

    // Decode and verify format: \0username\0password
    const encoded = authLine!.split(' ')[1]
    const decoded = Buffer.from(encoded, 'base64').toString('utf8')
    expect(decoded).toBe('\0myuser\0mypass')
  })

  /**
   * A password that is stored and cannot be read back is not a password.
   *
   * `decryptSecret` answers null both for "nothing saved" and for "saved under
   * a key this machine no longer has" — a moved profile, a reset keyring. The
   * second used to authenticate with an empty string, the server refused it,
   * and the banner said "SASL authentication failed": the server blamed for
   * turning down a password it was never sent, and no hint that the fix is to
   * type it again.
   */
  it('refuses to authenticate with a password it could not read', () => {
    const { client, sentLines } = createMockClient()
    client.config.saslMechanism = 'PLAIN'
    client.config.saslUsername = 'myuser'
    client.config.saslPassword = null
    client.config.unreadableSecrets = ['saslPassword']

    const errors: { message: string }[] = []
    client.events.on('error', (err: { message: string }) => errors.push(err))

    dispatchMessage(client, parseMessage('AUTHENTICATE +'))

    // Nothing that could be mistaken for an attempt
    const attempted = sentLines.filter(
      (l) => l.startsWith('AUTHENTICATE') && l !== 'AUTHENTICATE *'
    )
    expect(attempted).toEqual([])

    // The exchange is abandoned the way the spec says, and registration is
    // let go of — without CAP END the connection waits for ever.
    expect(sentLines).toContain('AUTHENTICATE *')
    expect(sentLines).toContain('CAP END')
    expect(client.state.capNegotiating).toBe(false)

    // And the message names the fix rather than the symptom
    expect(errors).toHaveLength(1)
    expect(errors[0].message).toMatch(/could not be read/i)
    expect(errors[0].message).toMatch(/enter it again/i)
  })

  /** A password that is simply absent is a different thing, and still tried */
  it('still authenticates when there is no password saved at all', () => {
    const { client, sentLines } = createMockClient()
    client.config.saslMechanism = 'PLAIN'
    client.config.saslUsername = 'myuser'
    client.config.saslPassword = null

    dispatchMessage(client, parseMessage('AUTHENTICATE +'))

    const authLine = sentLines.find((l) => l.startsWith('AUTHENTICATE') && l !== 'AUTHENTICATE +')
    expect(authLine).toBeDefined()
    expect(sentLines).not.toContain('AUTHENTICATE *')
  })

  it('handles SASL EXTERNAL — sends empty auth', () => {
    const { client, sentLines } = createMockClient()
    client.config.saslMechanism = 'EXTERNAL'

    dispatchMessage(client, parseMessage('AUTHENTICATE +'))

    expect(sentLines).toContain('AUTHENTICATE +')
  })

  it('sends CAP END on SASL success (903)', () => {
    const { client, sentLines, state } = createMockClient()

    dispatchMessage(client, parseMessage(':server 903 TestUser :SASL authentication successful'))

    expect(sentLines).toContain('CAP END')
    expect(state.capNegotiating).toBe(false)
  })

  it('sends CAP END on SASL failure (904)', () => {
    const { client, sentLines, state, events } = createMockClient()
    const handler = vi.fn()
    events.on('error', handler)

    dispatchMessage(client, parseMessage(':server 904 TestUser :SASL authentication failed'))

    expect(sentLines).toContain('CAP END')
    expect(state.capNegotiating).toBe(false)
    expect(handler).toHaveBeenCalledWith(
      expect.objectContaining({
        code: '904'
      })
    )
  })
})

File: android/app/src/test/java/org/switchboard/android/CapabilityTest.kt
package org.switchboard.android

import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.switchboard.android.irc.ConnectionState
import org.switchboard.android.irc.Handlers
import org.switchboard.android.irc.Irc
import org.switchboard.android.irc.IrcConnection
import org.switchboard.android.irc.IrcSession
import org.switchboard.android.irc.ServerConfig

/**
 * Capability negotiation on the phone.
 *
 * The desktop has had `capability.test.ts` and `capreq.test.ts` since it was
 * written; this is the phone's half. Getting any of it wrong fails silently:
 * only advertised capabilities are requested, so a wrong name is not an error
 * but an absence — the feature never turns on and nothing is logged. And
 * registration is held open until CAP END, so ending it a line early loses
 * whatever had not been answered, while never ending it hangs the connection
 * before 001.
 */
class CapabilityTest {

    private class Session(
        override val config: ServerConfig,
        override val state: ConnectionState
    ) : IrcSession {
        val sent = mutableListOf<String>()
        val events = mutableListOf<Pair<String, JsonObject>>()

        override fun send(command: String, vararg params: String) {
            sent.add(Irc.serialise(command, params.toList()))
        }

        override fun sendRaw(line: String) { sent.add(line) }
        override fun emit(channel: String, data: JsonElement) {
            events.add(channel to data.jsonObject)
        }
    }

    private lateinit var session: Session

    @Before
    fun setUp() {
        Handlers.installAll()
        val state = ConnectionState("srv")
        state.reset("kara")
        session = Session(
            ServerConfig(id = "srv", name = "Test", host = "irc.example.org", nick = "kara"),
            state
        )
    }

    private fun feed(line: String) = Handlers.dispatch(session, Irc.parse(line))

    /** The capabilities asked for, in the order they were asked for */
    private fun requested(): List<String> = session.sent
        .filter { it.startsWith("CAP REQ") }
        .flatMap { it.removePrefix("CAP REQ ").removePrefix(":").split(" ") }
        .filter { it.isNotEmpty() }

    // ── what we ask for ───────────────────────────────────────────────

    @Test
    fun `asks only for what the server advertised`() {
        feed(":irc.example.org CAP * LS :multi-prefix server-time nonsense-cap")

        // In our own order of preference, not the server's order of listing —
        // the same order the desktop asks in, which is what makes the two
        // negotiate identically against the same network
        assertEquals(listOf("server-time", "multi-prefix"), requested())
    }

    @Test
    fun `keeps a capability's value without asking for it`() {
        feed(":irc.example.org CAP * LS :sasl=PLAIN,EXTERNAL server-time")

        assertEquals("PLAIN,EXTERNAL", session.state.available["sasl"])
        assertTrue("the name is asked for, not the name and its value",
            requested().contains("sasl"))
    }

    @Test
    fun `waits for the rest of a multi-line listing before asking`() {
        feed(":irc.example.org CAP * LS * :multi-prefix")
        assertTrue("nothing is requested until the last line", requested().isEmpty())

        feed(":irc.example.org CAP * LS :server-time")
        assertEquals(listOf("server-time", "multi-prefix"), requested())
    }

    @Test
    fun `ends negotiation at once when the server offers nothing we want`() {
        feed(":irc.example.org CAP * LS :nonsense-cap")

        assertTrue(requested().isEmpty())
        assertEquals(listOf("CAP END"), session.sent)
    }

    // ── splitting the request ─────────────────────────────────────────

    @Test
    fun `splits a long list across lines that each fit`() {
        // Everything the client knows how to ask for, which is far more than
        // one 512-byte line holds
        val all = IrcConnection.WANTED_CAPABILITIES
        feed(":irc.example.org CAP * LS :" + all.joinToString(" "))

        val lines = session.sent.filter { it.startsWith("CAP REQ") }
        assertTrue("a list this long needs more than one line", lines.size > 1)
        for (line in lines) {
            assertTrue(
                "a CAP REQ has to fit in one message: ${line.length}",
                line.toByteArray().size + 2 <= IrcConnection.MAX_LINE_BYTES
            )
        }
    }

    @Test
    fun `asks for everything, exactly once, in order`() {
        val all = IrcConnection.WANTED_CAPABILITIES
        feed(":irc.example.org CAP * LS :" + all.joinToString(" "))

        assertEquals(all, requested())
    }

    @Test
    fun `counts the lines, so CAP END waits for the last answer`() {
        val all = IrcConnection.WANTED_CAPABILITIES
        feed(":irc.example.org CAP * LS :" + all.joinToString(" "))

        val lines = session.sent.filter { it.startsWith("CAP REQ") }
        assertEquals(lines.size, session.state.pendingCapRequests)

        // Answer all but the last: registration must stay open
        for (line in lines.dropLast(1)) {
            feed(":irc.example.org CAP kara ACK :" + line.substringAfter(":"))
        }
        assertFalse("CAP END must not go early", session.sent.contains("CAP END"))

        feed(":irc.example.org CAP kara ACK :" + lines.last().substringAfter(":"))
        assertTrue("and must go once the last is answered", session.sent.contains("CAP END"))
    }

    // ── the answers ───────────────────────────────────────────────────

    @Test
    fun `an ACK records what was granted and ends negotiation`() {
        feed(":irc.example.org CAP * LS :multi-prefix server-time")
        feed(":irc.example.org CAP kara ACK :multi-prefix server-time")

        assertTrue(session.state.capabilities.contains("multi-prefix"))
        assertTrue(session.state.capabilities.contains("server-time"))
        assertTrue(session.sent.contains("CAP END"))
    }

    @Test
    fun `a NAK ends negotiation too, rather than waiting for ever`() {
        feed(":irc.example.org CAP * LS :multi-prefix")
        feed(":irc.example.org CAP kara NAK :multi-prefix")

        assertFalse("nothing was granted", session.state.capabilities.contains("multi-prefix"))
        assertTrue("but registration still has to finish", session.sent.contains("CAP END"))
    }

    @Test
    fun `CAP NEW asks for something the server has only just offered`() {
        feed(":irc.example.org CAP * LS :multi-prefix")
        feed(":irc.example.org CAP kara ACK :multi-prefix")
        session.sent.clear()

        feed(":irc.example.org CAP kara NEW :server-time")

        assertEquals(listOf("server-time"), requested())
        assertFalse(
            "a second CAP END would end a registration that is already over",
            session.sent.contains("CAP END")
        )
    }

    @Test
    fun `CAP DEL takes one away again`() {
        feed(":irc.example.org CAP * LS :multi-prefix server-time")
        feed(":irc.example.org CAP kara ACK :multi-prefix server-time")

        feed(":irc.example.org CAP kara DEL :server-time")

        assertTrue(session.state.capabilities.contains("multi-prefix"))
        assertFalse(session.state.capabilities.contains("server-time"))
    }

    // ── SASL holds the door open ──────────────────────────────────────

    @Test
    fun `with SASL to do, CAP END waits for the exchange`() {
        val withSasl = Session(
            session.config.copy(saslMechanism = "PLAIN", saslPassword = "hunter2"),
            session.state
        )
        Handlers.dispatch(withSasl, Irc.parse(":irc.example.org CAP * LS :sasl=PLAIN"))
        Handlers.dispatch(withSasl, Irc.parse(":irc.example.org CAP kara ACK :sasl"))

        assertTrue("authentication starts", withSasl.sent.any { it.startsWith("AUTHENTICATE") })
        assertFalse(
            "and CAP END waits for it, or the account arrives after the nick",
            withSasl.sent.contains("CAP END")
        )
    }

    @Test
    fun `a mechanism the server does not offer is said out loud, not left to fail`() {
        val withSasl = Session(
            session.config.copy(saslMechanism = "SCRAM-SHA-512", saslPassword = "hunter2"),
            session.state
        )
        Handlers.dispatch(withSasl, Irc.parse(":irc.example.org CAP * LS :sasl=PLAIN,EXTERNAL"))
        Handlers.dispatch(withSasl, Irc.parse(":irc.example.org CAP kara ACK :sasl"))

        val errors = withSasl.events.filter { it.first == "irc:error" }
        assertTrue("a bare 904 tells the user their password is wrong", errors.isNotEmpty())
        assertTrue(
            "and says what the server does take",
            errors.any { it.second.toString().contains("PLAIN") }
        )
        assertTrue("registration still finishes", withSasl.sent.contains("CAP END"))
    }
}
